ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -287,12 +305,143 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
+	if rayJobInstance.Spec.SubmitterPodTemplate == nil {
+		// Set the default value
+		submitter_template = v1.PodTemplateSpec{
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  "ray-job-submitter",
+						Image: "rayproject/ray:2.5.0",
+						Resources: v1.ResourceRequirements{
+							Limits: v1.ResourceList{
+								v1.ResourceCPU: resource.MustParse("1"),
+							},
+							Requests: v1.ResourceList{
+								v1.ResourceCPU: resource.MustParse("200m"),
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+			},
+		}
+	} else {
+		submitter_template = *rayJobInstance.Spec.SubmitterPodTemplate.DeepCopy()
+	}
+
+	address := rayJobInstance.Status.DashboardURL
+
+	// add http:// if needed
+	if !strings.HasPrefix(address, "http://") {
+		address = "http://" + address
+	}
+
+	k8s_job_command := []string{"ray", "job", "submit", "--address", address}
+
+	// Add runtimeEnv to command
+	if len(rayJobInstance.Spec.RuntimeEnv) > 0 {
+		decodedBytes, err := base64.StdEncoding.DecodeString(rayJobInstance.Spec.RuntimeEnv)
+		if err != nil {
+			return "", fmt.Errorf("failed to decode runtimeEnv: %v: %v", rayJobInstance.Spec.RuntimeEnv, err)
+		}
+		runtimeEnv := string(decodedBytes)
+		k8s_job_command = append(k8s_job_command, "--runtime-env-json", runtimeEnv)
+	}
+
+	// Add metadata to command
+	if len(rayJobInstance.Spec.Metadata) > 0 {
+		// Check that the Ray version is at least 2.6.0.
+		// If it is, we can use the --metadata-json flag.
+		// Otherwise, we need to raise an error.
+
+		rayVersion := rayJobInstance.Spec.RayClusterSpec.RayVersion
+		constraint, _ := semver.NewConstraint(">= 2.6.0")
+		v, err := semver.NewVersion(rayVersion)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse Ray version: %v: %v", rayVersion, err)
+		}
+
+		if !constraint.Check(v) {
+			return "", fmt.Errorf("the Ray version must be at least 2.6.0 to use the metadata field")
+		}
+
+		// Convert the metadata map to a JSON string.
+		metadataBytes, err := json.Marshal(rayJobInstance.Spec.Metadata)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal metadata: %v: %v", rayJobInstance.Spec.Metadata, err)
+		}
+		metadata := string(metadataBytes)
+
+		k8s_job_command = append(k8s_job_command, "--metadata-json", metadata)
+	}
+
+	// Add jobid to command
+	if len(rayJobInstance.Status.JobId) > 0 {
+		jobid := rayJobInstance.Status.JobId
+		k8s_job_command = append(k8s_job_command, "--job-id", jobid)
+	}
+
+	entrypoint := rayJobInstance.Spec.Entrypoint
+	commandSlice, err := shlex.Split(entrypoint)
+	if err != nil {
+		return "", err
+	}
+
+	submitter_template.Spec.Containers[0].Command = append(k8s_job_command, commandSlice...)
+
+	jobName := rayJobInstance.Name
+	jobNamespace := rayJobInstance.Namespace
+
+	// Create a Job object with the specified name and namespace
+	job := &batchv1.Job{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: jobNamespace, Name: jobName}, job); err != nil {
```

Comment:
Maybe we should check this at first before setting up `submitter_template`?

Reply:
Fixed 4b6f489